### ISSUE TITLE
JWT Populate Lambda doesn't run on authenticationType REGISTRATION

### ISSUE THREAD

**Issue by calebfreeman** [FusionAuth/fusionauth-issues#856]
### Description 
I'm using Hasura and need certain claims to populate in the JWT else the frontend will throw an error. I have a JWT Populate Lambda setup in FusionAuth to do this. The token returned in the /api/user/registration/ api doesn't get populated with that Lambda. I've looked around for a setting to turn this on, or documentation describing why this may happen. I haven't been able to find it. It would seem that the access token returned in the registration API should populate just like any other access token. To get around this I am ignoring the access token returned in the registration api and calling FusionAuth again with the refresh token returned in the registration call to retrieve another access token which populates using the lambda.

### Affects versions
1.18.4

### Steps to reproduce
1. Create/configure application that uses a JWT Populate Lambda that adds claims to the JWT
2. Register user to that application using FusionAuth api
3. Introspect the returned access token and see the claims from the Lambda are not there

### Expected behavior
The access token returned from the /api/user/registration/ api should be populated with the JWT Populate Lambda, if that is not expected behavior then it should be documented.

### Platform
 - Device: MacBook Pro
 - OS: MacOS 10.14 (running FusionAuth on Docker)

**Comment by robotdan** [comment]
Can you try a few things? 

1. Add at least one log statement to create a JWT populate event log. This will tell us if the lambda is being run.  For example, `console.info('here!');`
2. Verify the JWT Populate lambda is enabled. `Application > Edit > JWT > Access Token populate lambda` .
3. Take the encoded JWT returned on the Registration API and paste it in this JWT debugger to ensure the claims are expected. https://fusionauth.io/learn/expert-advice/dev-tools/jwt-debugger

**Comment by calebfreeman** [comment]
I identified the issue. FusionAuth does call Lambda and looks for the user's roles, but throws an error because it cannot find them:

TypeError: Cannot get property "roles" of null in <eval> at line number 3

This is what me JWT Populate Lambda looks like:

function populate(jwt, user, registration) {
  jwt['https://hasura.io/jwt/claims'] = {
    "x-hasura-allowed-roles": registration.roles,
    "x-hasura-default-role": "anonymous",
    "x-hasura-user-id": user.id
  };
}

I'm looking for registration.roles but registration doesn't exist. I'm expecting registration.roles to be an array of strings. Is this expected behavior on registration?

The JWT I get after the failed JWT Populate is this:

{
  "aud": "5f36d9bd-7344-42e0-80d9-50458f17fbd4",
  "exp": 1599838716,
  "iat": 1599835116,
  "iss": "domain.dev",
  "sub": "8391706b-9c5d-4a7d-bda9-0ed4d84b3ef0",
  "jti": "e832e5af-3d82-4517-8d04-b1273d5daa1f",
  "authenticationType": "REGISTRATION",
  "email": "email@email.com",
  "email_verified": false,
  "preferred_username": "username",
  "applicationId": "5f36d9bd-7344-42e0-80d9-50458f17fbd4",
  "roles": [
    {
      "id": "a87e53c0-cf51-4d61-9d93-da4548333edd",
      "insertInstant": 1597800913,
      "isDefault": true,
      "isSuperRole": false,
      "lastUpdateInstant": 1597800913,
      "name": "anonymous"
    },
    {
      "id": "c1ae6c38-f73e-4b9c-93c1-c674223a9626",
      "insertInstant": 1597800105,
      "isDefault": false,
      "isSuperRole": false,
      "lastUpdateInstant": 1597800922,
      "name": "user"
    }
  ]
}

**Comment by calebfreeman** [comment]
below are the objects available in the JWT populate function on registration. To get the roles array of strings I would need to iterate through the roles object in the jwt object. This is on the combined create/register user api. Fusionauth correctly registers the user to my application. Seems like it should have the registration object available. 

user:
```
{"active":true,"connectorId":"e3306678-a53a-4964-9040-1c96f36dda72",
"data":{},"email":"email",
"firstName":"fn",
"fullName":"full name",
"id":"4c06da62-b906-4131-be8c-154a06fc2df1",
"insertInstant":1599836547448,
"lastLoginInstant":1599836547663,
"lastName":"ln",
"lastUpdateInstant":1599836547448,
"memberships":[],
"middleName":"N",
"mobilePhone":"0000000000",
"passwordChangeRequired":false,
"passwordLastUpdateInstant":1599836547474,
"preferredLanguages":[],
"registrations":[],
"tenantId":"ea2099c0-dab9-4b22-9190-5e1872c820a4",
"timezone":"America/New_York",
"twoFactorDelivery":"None",
"twoFactorEnabled":false,
"username":"0000000000",
"usernameStatus":"ACTIVE",
"verified":false}
```
registration:
`null`
jwt:
```
{"applicationId":"5f36d9bd-7344-42e0-80d9-50458f17fbd4",
"aud":"5f36d9bd-7344-42e0-80d9-50458f17fbd4",
"authenticationType":"REGISTRATION",
"email":"email",
"email_verified":false,
"exp":1599840147,
"iat":1599836547,
"iss":"domain.dev",
"jti":"d461c28f-46e1-4662-a6db-fb7864bf9749",
"preferred_username":"username",
"roles":[{"id":"a87e53c0-cf51-4d61-9d93-da4548333edd",
"insertInstant":1597800913802,
"isDefault":true,
"isSuperRole":false,
"lastUpdateInstant":1597800913802,
"name":"anonymous"
},{"id":"c1ae6c38-f73e-4b9c-93c1-c674223a9626",
"insertInstant":1597800105579,
"isDefault":false,
"isSuperRole":false,
"lastUpdateInstant":1597800922383,
"name":"user"}],
"sub":"4c06da62-b906-4131-be8c-154a06fc2df1",
"https://hasura.io/jwt/claims":{"x-hasura-allowed-roles":["user","anonymous"],
"x-hasura-default-role":"anonymous",
"x-hasura-user-id":"4c06da62-b906-4131-be8c-154a06fc2df1"}
}
```

**Comment by mooreds** [comment]
Hiya,

Glad you were able to get the lambda to run. I'm working off of the docs here: https://hasura.io/docs/1.0/graphql/core/auth/authentication/jwt.html (which I believe is the current version). Here's the lambda I'd use:

```
function populate(jwt, user, registration) {
  var roles = [];
  if (registration && registration.roles) {
    roles = registration.roles;
  }
  console.log("user.id: "+user.id);
  console.log("roles for user: " + roles);
  jwt['https://hasura.io/jwt/claims'] = {
    "x-hasura-allowed-roles": roles,
    "x-hasura-default-role": "anonymous",
    "x-hasura-user-id": user.id
  };
  console.log("got here");
}
```

The fact that `registration` is null indicates to me that perhaps you didn't register this user with an application? Roles are assigned to a user who is registered with an application.

**Comment by mooreds** [comment]
Ah wait! I see that you are doing a registration. Let me dive in a bit more.

**Comment by mooreds** [comment]
Yes, it sure looks to me like the registration object isn't being passed to the lambda when I create a user via the API. It does end up in the JWT in the roles object, but isn't in the lambda.

To recreate:

* Create an app, add a role `test role` to it.
* Create a lambda:
```
function populate(jwt, user, registration) {
  console.log("registration: "+registration);
  var roles = [];
  if (registration && registration.roles) {
    roles = registration.roles;
  }
  console.log("user.id: "+user.id);
  console.log("roles for user: " + roles);
  jwt['https://hasura.io/jwt/claims'] = {
    "x-hasura-allowed-roles": roles,
    "x-hasura-default-role": "anonymous",
    "x-hasura-user-id": user.id
  };
  console.log("got here");
}
```
* Associate this lambda with the application
* Run this command (updating the path and application id if needed)
```
curl -vvvv -XPOST -H 'Content-type: application/json' -H "Authorization: $API_KEY" 'http://localhost:9011/api/user/registration' -d '{"user": {"email": "testit13@example.com", "password": "password" }, "registration" : { "applicationId" : "e69d45b2-d3f5-4fcc-b35d-81ee338b027e", "roles" : [ "test role" ] }}'
```

Go to the event log, you see: 

```
Lambda invocation result.

Id: 18cca09b-feea-4de3-9566-41f57ef3ad26
Name: testpopulate

registration: null
user.id: 72764427-bea9-4302-9303-372800923ab6
roles for user: 
got here
```

This was with version 1.18.7.

**Comment by robotdan** [comment]
This is a bug. It will be fixed in the next patch release.

**Comment by robotdan** [comment]
Fix available in version 1.19.4.
